### PR TITLE
[SPARK-50125] Add new error class - Data source does not support DML operation

### PR DESCRIPTION
**Pull request by urosstan-db**
### What changes were proposed in this pull request?
Add new error class which indicates certain data source does not support DML operation.

### Why are the changes needed?
Sometimes, Data Source V2 does not support DML operations such as create table, alter table, drop namespace, etc i.e. it is read only data source, so we need some new error class which will be thrown in case we implement such data source.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
No testing needed, compile only

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: common/utils/src/main/resources/error/error-conditions.json [hunk 1]
```diff
@@ -1091,6 +1091,12 @@
     ],
     "sqlState" : "KD010"
   },
+  "DATA_SOURCE_UNSUPPORTED_DML_OPERATION" : {
```
Comment (urosstan-db) [inline common/utils/src/main/resources/error/error-conditions.json:1094]: Maybe subclasses for each operation is better?
Operations are something like
'CREATE TABLE', 'ALTER NAMESPACE', etc

File: common/utils/src/main/resources/error/error-conditions.json [hunk 2]
```diff
@@ -1091,6 +1091,12 @@
...
+    "message" : [
+      "Data source '<provider>' does not support requested DML operation: '<operation>'"
+    ],
+    "sqlState" : "KD011"
+  },
   "DATA_SOURCE_NOT_EXIST" : {
     "message" : [
```
Comment (urosstan-db) [inline common/utils/src/main/resources/error/error-conditions.json:1101]: I will add JIRA item when we are satisfied with the changes in PR

File: common/utils/src/main/resources/error/error-conditions.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (urosstan-db) [reply common/utils/src/main/resources/error/error-conditions.json:1094]: This error class should be for data sources which does not have implementation for some DML operation, not like transient state.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 3]
```diff
@@ -1091,6 +1091,12 @@
     ],
     "sqlState" : "KD010"
   },
+  "DATA_SOURCE_UNSUPPORTED_DML_OPERATION" : {
+    "message" : [
+      "Data source '<provider>' does not support requested DML operation: '<operation>'"
```
[the remark this answers is not in the document]
Comment (urosstan-db) [reply common/utils/src/main/resources/error/error-conditions.json:1096]: Thanks, TBD

[the remark this answers is not in the document]
Comment (urosstan-db) [reply common/utils/src/main/resources/error/error-conditions.json:1096]: Even for provider as well? Or just for SQL operation?

[the remark this answers is not in the document]
Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-conditions.json:1096]: for the dml operation. Are you going to pass some SQL ops there, correct?

[the remark this answers is not in the document]
Comment (urosstan-db) [reply common/utils/src/main/resources/error/error-conditions.json:1096]: Yes, some descriptive SQL operations, e.g. CREATE TABLE, DROP NAMESPACE, etc.

[the remark this answers is not in the document]
Comment (urosstan-db) [reply common/utils/src/main/resources/error/error-conditions.json]: Done, I also renamed `operation` to `sqlOperation`

File: common/utils/src/main/resources/error/error-conditions.json (hunk 1 above)
Comment (srielau) [reply common/utils/src/main/resources/error/error-conditions.json:1094]: It is common to name the operation: "ALTER TABLE', "UPDATE" as a parameter.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 4]
```diff
@@ -1091,6 +1091,12 @@
     ],
     "sqlState" : "KD010"
   },
+  "DATA_SOURCE_UNSUPPORTED_DML_OPERATION" : {
+    "message" : [
+      "Data source '<provider>' does not support requested DML operation: <sqlOperation>"
+    ],
+    "sqlState" : "KD011"
```
Comment (srielau) [inline common/utils/src/main/resources/error/error-conditions.json:1098]: You need to propose a KD011, it seems new.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 5]
```diff
@@ -1091,6 +1091,12 @@
     ],
     "sqlState" : "KD010"
   },
+  "DATA_SOURCE_UNSUPPORTED_DML_OPERATION" : {
+    "message" : [
+      "Data source '<provider>' does not support requested DML operation: <sqlOperation>"
```
Comment (srielau) [inline common/utils/src/main/resources/error/error-conditions.json:1096]:
```suggestion
      "Data source '<provider>' does not support requested operation: <sqlOperation>"
```
We can be more generic.

File: common/utils/src/main/resources/error/error-conditions.json (hunk 4 above)
Comment (urosstan-db) [reply common/utils/src/main/resources/error/error-conditions.json:1098]: Done, thanks

File: common/utils/src/main/resources/error/error-conditions.json (hunk 5 above)
Comment (urosstan-db) [reply common/utils/src/main/resources/error/error-conditions.json]: Done

File: common/utils/src/main/resources/error/error-states.json [hunk 6]
```diff
@@ -7441,6 +7441,12 @@
         "standard": "N",
         "usedBy": ["Databricks"]
     },
+    "KD011": {
+        "description": "external data source does not support DML operation",
```
Comment (srielau) [inline common/utils/src/main/resources/error/error-states.json:7445]:
```suggestion
        "description": "external data source does not support operation",
```

**Review by srielau — APPROVED** [review]

File: common/utils/src/main/resources/error/error-conditions.json [hunk 7]
```diff
@@ -1111,6 +1111,12 @@
     ],
     "sqlState" : "42K03"
   },
+  "DATA_SOURCE_UNSUPPORTED_DML_OPERATION" : {
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-conditions.json:1114]: if we removed DML from message, let's remove it from the condition name.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 8]
```diff
@@ -1111,6 +1111,12 @@
...
+    "message" : [
+      "Data source '<provider>' does not support requested operation: <operation>"
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-conditions.json:1116]:
```suggestion
      "Data source '<provider>' does not support the requested operation: <operation>"
```

**Review by MaxGekk — COMMENTED** [review]
@urosstan-db Are you going to raise the error from somewhere?

File: common/utils/src/main/resources/error/error-conditions.json (hunk 7 above)
Comment (urosstan-db) [reply common/utils/src/main/resources/error/error-conditions.json:1114]: I want to have it just exception dedicated to DML operations, if it is needed I can revert DML in message? Serge suggested me to do that

**Comment by urosstan-db** [comment]
> @urosstan-db Are you going to raise the error from somewhere?

No for now, not in Spark to be more precise.

**Comment by MaxGekk** [comment]
> No for now, not in Spark to be more precise.

In that case, I don't think it should be merged to OSS Spark, sorry.

File: common/utils/src/main/resources/error/error-conditions.json (hunk 7 above)
Comment (srielau) [reply common/utils/src/main/resources/error/error-conditions.json:1114]: The message should be in sync with the name. So yes, I implied to remove it from both.

**Comment by urosstan-db** [comment]
> > No for now, not in Spark to be more precise.
> 
> In that case, I don't think it should be merged to OSS Spark, sorry.

Ok thanks, I thought to avoid introducing of multiple similar codes, but let's close this one then.

**Pull request closed without merging**